File: app/src/main/java/rx_playground/com/jablonski/rxandroidplayground/submodels/SubmodelsListPresenter.java
package rx_playground.com.jablonski.rxandroidplayground.submodels;

import java.util.List;

import rx_playground.com.jablonski.rxandroidplayground.mvp.BaseViewCotract;
import rx_playground.com.jablonski.rxandroidplayground.model.Model;

/**
 * Created by yabol on 23.04.2017.
 */

public class SubmodelsListPresenter implements SubmodelsViewContract.Presenter, BaseViewCotract.BaseProvider<Model>, BaseViewCotract.BaseOnItemCLickListener<Model> {
    private SubmodelsViewContract.View view;
    private SubmodelsViewContract.Repository repository;
    private List<Model> submodels;

    public SubmodelsListPresenter(SubmodelsViewContract.View view){
        this.view = view;
    }

    public void setRepository(SubmodelsViewContract.Repository repository){
        this.repository = repository;
    }
    @Override
    public void displayElements(List<Model> elements) {
        this.submodels = elements;
        this.view.showView(elements);
    }

    @Override
    public List<Model> getElements() {
        return this.submodels;
    }

    @Override
    public void loadElements(String manufacturer, String niceName, String year) {
        this.repository.loadElements(manufacturer, niceName, year);
    }

    @Override
    public Model getObject(int position) {
        if(this.submodels != null) {
            return this.submodels.get(position);
        }
        return null;
    }

    @Override
    public int getCount() {
        if(this.submodels != null){
            return this.submodels.size();
        }
        return 0;
    }

    @Override
    public void performClick(Model object) {
        view.openModelDetails(object.getId());
    }
}

File: app/src/main/java/rx_playground/com/jablonski/rxandroidplayground/submodels/SubmodelsListPresenter.kt
package rx_playground.com.jablonski.rxandroidplayground.submodels

import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import rx_playground.com.jablonski.rxandroidplayground.model.Model
import rx_playground.com.jablonski.rxandroidplayground.model.ModelsResult
import rx_playground.com.jablonski.rxandroidplayground.mvp.BaseViewCotract

/**
 * Created by yabol on 23.04.2017.
 */

class SubmodelsListPresenter(private val view: SubmodelsViewContract.View, private var repository: SubmodelsViewContract.Repository) : SubmodelsViewContract.Presenter, BaseViewCotract.BaseProvider<Model>, BaseViewCotract.BaseOnItemCLickListener<Model> {
    private var submodels: List<Model>? = null

    override fun displayElements(elements: List<Model>?) {
        this.submodels = elements
        this.view.showView(elements)
    }

    override fun getElements(): List<Model>? {
        return this.submodels
    }

    override fun loadElements(manufacturer: String, niceName: String, year: String) {
        var elements : List<Model>? = null
        this.repository.loadElements(manufacturer, niceName, year).subscribe(object : Observer<ModelsResult> {
            override fun onSubscribe(d: Disposable) {

            }

            override fun onNext(modelsResult: ModelsResult) {
                if (modelsResult.models != null && modelsResult.models.size > 0) {
                    elements = modelsResult.models[0].submodels
                }
            }

            override fun onError(e: Throwable) {

            }

            override fun onComplete() {
                elements?.let {
                    displayElements(it)
                }

            }
        })
    }

    override fun getObject(position: Int): Model? {
        return if (this.submodels != null) {
            this.submodels?.let {
                it[position]
            }
        } else null
    }

    override fun getCount(): Int {
        return if (this.submodels != null) {
            this.submodels!!.size
        } else 0
    }

    override fun performClick(`object`: Model) {
        view.openModelDetails(`object`.id)
    }
}
